be/src/fs/hdfs/hdfs_fs_cache.cpp

```diff
@@ -69,32 +68,42 @@ static Status create_hdfs_fs_handle(const std::string& namenode, HdfsFsHandle* h
 }
 
 Status HdfsFsCache::get_connection(const std::string& namenode, HdfsFsHandle* handle, const FSOptions& options) {
-    {
-        std::lock_guard<std::mutex> l(_lock);
-        std::string cache_key = namenode;
-        const THdfsProperties* properties = options.hdfs_properties();
-        if (properties != nullptr && properties->__isset.hdfs_username) {
-            cache_key += properties->hdfs_username;
-        }
+    std::lock_guard<std::mutex> l(_lock);
+    std::string cache_key = namenode;
+    const THdfsProperties* properties = options.hdfs_properties();
+    if (properties != nullptr && properties->__isset.hdfs_username) {
+        cache_key += properties->hdfs_username;
+    }
 
-        // Insert azure cloud credential into cache key
-        const std::vector<TCloudProperty>* azure_cloud_properties = get_azure_cloud_properties(options);
-        if (azure_cloud_properties != nullptr) {
-            for (const auto& cloud_property : *azure_cloud_properties) {
-                cache_key += cloud_property.key;
-                cache_key += cloud_property.value;
-            }
+    // Insert azure cloud credential into cache key
+    const std::vector<TCloudProperty>* azure_cloud_properties = get_azure_cloud_properties(options);
+    if (azure_cloud_properties != nullptr) {
+        for (const auto& cloud_property : *azure_cloud_properties) {
+            cache_key += cloud_property.key;
+            cache_key += cloud_property.value;
         }
+    }
 
-        auto it = _cache.find(cache_key);
-        if (it != _cache.end()) {
-            *handle = it->second;
-        } else {
-            handle->namenode = namenode;
-            RETURN_IF_ERROR(create_hdfs_fs_handle(namenode, handle, options));
-            _cache[cache_key] = *handle;
+    for(size_t idx = 0; idx < _cur_client_idx; idx++) {
+        if (_cache_key[idx] == cache_key) {
+            *handle = _cache_clients[idx];
+            // Found cache client, return directly
+            return Status::OK();
         }
     }
+
+    // Not found cached client, create a new one
+    handle->namenode = namenode;
+    RETURN_IF_ERROR(create_hdfs_fs_handle(namenode, handle, options));
+    if (UNLIKELY(_cur_client_idx >= _max_cache_clients)) {
+        uint32_t idx = _rand.Uniform(_max_cache_clients);
```

Comment:
We should call hdfsDisconnect on the evicted client

> Connect to HDFS using the parameters defined by the builder. The HDFS builder will be freed, whether or not the connection was successful. Every successful call to hdfsBuilderConnect should be matched with a call to hdfsDisconnect, when the hdfsFS is no longer needed.

Reply:
There are chances that another thread is using the evicted client. 

We could use the shared pointer instead of the raw pointer, just like `fs_s3.cpp`

Reply:
Great point.

Reply:
> There are chances that another thread is using the evicted client.
> 
> We could use the shared pointer instead of the raw pointer, just like `fs_s3.cpp`

We have using `std::lock_guard<std::mutex> l(_lock);` to avoid race.

Reply:
The parameter `handle` is not guarded by the mutex once returned. And `handle->hdfs_fs`, which is of type `hdfs_internal*`, may point to the same address with some cached handle.

Reply:
Changed, using `shared_ptr` instead.